### PR TITLE
Fix compiler hang with recursive function

### LINKED ISSUE #11111: Compiler hang with recursive function passing struct pointer to itself
# Issue Description

Observe the reproducer code below if the title is unclear. This requires a CPU target to reproduce, as other targets do not support recursion. The hang occurs in the `lower-buffer-element-type` pass.

This specifically requires both of these conditions:
* A recursive function with an `inout` parameter, passing the parameter to itself
* The type of the parameter must be a `struct`

# Reproducer Code
```slang
struct Foo
{
    int f;
}

void recursive(Foo* i)
{
    recursive(i);
}

export __extern_cpp int main(int argc, NativeString* argv)
{
    Foo* i; // This could also originate from a valid allocation, that also causes the exact same behavior.
    recursive(i);
    return 0;
}
```
(compile with `slangc -target llvm-host-ir -emit-cpu-via-llvm test.slang -o test.ll`)

# Expected Behavior
Compiler finishes compiling eventually.

# Actual Behavior
Compiler hangs eternally, stuck looping in a never-emptying work list in `deferStorageToLogicalCasts`.

# Environment
 - Slang Version eab26890
 - OS: OpenSUSE Tumbleweed

### PR DESCRIPTION
**Pull request by juliusikkala**
Fixes #11111.

### CODE REVIEW

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 1]
```diff
@@ -1778,7 +1778,13 @@ struct LoweredElementTypeContext
         // This means that `FieldAddr(CastStorageToLogical(buffer), field0))` is translated to
         // `CastStorageToLogical(FieldAddr(buffer, field0))`. This way we can be sure that we are
         // doing minimal packing/unpacking.
-        deferStorageToLogicalCasts(module, _Move(castInstWorkList));
+        //
+        // This operation is problematic on recursive functions, which are
+        // allowed by CPU targets. Luckily, such targets also have downstream
+        // compilers with further optimizations, rendering this pass unnecessary
+        // there.
+        if (!isCPUTarget(target->getTargetReq()))
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp]: I do not think this optimization is doing anything of value on CPUs, as they have very proficient optimizers on their respective downstream compilers. So yes, this is a conservative choice, but also a "save compilation time on targets that don't need this" choice.

However, if the tests fail, then this pass is needed for validity too. In that case, I'll reconsider this approach.

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 2]
```diff
@@ -1778,7 +1778,13 @@ struct LoweredElementTypeContext
...
+            deferStorageToLogicalCasts(module, _Move(castInstWorkList));
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp]: Unfortunately that check won't work, as the function pointer changes on every iteration as it gets copied by that very pass. I would've done it that way if it was so easy.

[the remark this answers is not in the document]
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp]: pointers count as buffer types in this context

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 3]
```diff
@@ -1585,6 +1585,30 @@ struct LoweredElementTypeContext
         auto clonedFunc = as<IRFunc>(cloneInst(&cloneEnv, &builder, call->getCallee()));
         List<IRUse*> uses;
 
+        // If our cloned function has uses already, that means it's calling
+        // itself (= is recursive). Such uses are problematic, as the
+        // `specializedFunc` cache won't contain the newly cloned function,
+        // causing an infinite loop of specialization. The original has already
+        // been handled though, so we can just swap the calls back to calling
+        // the original function and avoid such loops.
+        if (clonedFunc->hasUses())
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp:1613]: That's actually not legal, Slang does not have function pointers. But, I can gate the IRCall update with `user->getCallee() == clonedFunc`, shouldn't do any harm.

File: tests/llvm/recursion-hang.slang [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+{
+    recursive(i);
+}
+
+export __extern_cpp int main(int argc, NativeString* argv)
+{
+    Foo* i = nullptr;
+    recursive(i);
+    return 0;
+}
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply tests/llvm/recursion-hang.slang:22]: Mutual recursion is not an issue already, and conditional recursion is a pointless extra test that doesn't exercise anything interesting. The CHECK would be uninteresting in any case; if this regresses, the compiler WILL hang. Making it a CHECK test doesn't make it not hang.

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 5]
```diff
@@ -1585,6 +1585,30 @@ struct LoweredElementTypeContext
...
+        {
+            // Replace all self-calls with the original function. This cuts
+            // the loop, as `specializedFunc` cache contains the original
+            // function now.
+            cloneEnv.mapOldValToNew.remove(clonedFunc);
+            cloneEnv.mapOldValToNew.add(clonedFunc, call->getCallee());
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp:1617]: The `remove` cannot be removed, as `add` asserts if the key already exists. Which in this case it does.

File: tests/llvm/recursion-hang.slang [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// This test is LLVM only due to it being the only backend where:
+// * recursion is allowed
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply tests/llvm/recursion-hang.slang:2]: It doesn't claim that, it claims that there are two conditions and that's one of the two.

File: tests/llvm/recursion-hang.slang [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+// * this bug manifests
+//TEST:SIMPLE: -target llvm-host-ir -emit-cpu-via-llvm
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply tests/llvm/recursion-hang.slang:4]: I know, right. We can't get the output because the compiler hangs :roll_eyes:

File: tests/llvm/recursion-hang.slang [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+struct Foo
+{
+    int f;
+}
+
+// This used to cause a compiler hang due to the lowerBufferElementTypes pass
+// deferring the pointer load infinitely deep in the callstack.
+void recursive(Foo* i)
+{
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply tests/llvm/recursion-hang.slang:13]: The other scenarios aren't bugged already, they don't suffer from this corner case.

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 9]
```diff
@@ -1585,6 +1585,34 @@ struct LoweredElementTypeContext
         auto clonedFunc = as<IRFunc>(cloneInst(&cloneEnv, &builder, call->getCallee()));
         List<IRUse*> uses;
 
+        // If our cloned function has uses already, that means it's calling
+        // itself (= is recursive). Such uses are problematic, as the
+        // `specializedFunc` cache won't contain the newly cloned function,
+        // causing an infinite loop of specialization. The original has already
+        // been handled though, so we can just swap the calls back to calling
+        // the original function and avoid such loops.
+        if (clonedFunc->hasUses())
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp:1613]: It's not 100% perfect, but because there are no function pointers in Slang, I think it's pretty close. Either way, this gate is only for "optimization" anyway; the loop below does nothing of significance if there are no uses of the cloned function.

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 10]
```diff
@@ -1585,6 +1585,34 @@ struct LoweredElementTypeContext
...
+                [&](IRCall* user)
+                {
+                    // Ensure that the use is specifically as the callee.
+                    if (user->getCallee() == clonedFunc)
+                    {
+                        builder.setInsertBefore(user);
+                        auto newCall = cloneInst(&cloneEnv, &builder, user);
+                        user->replaceUsesWith(newCall);
+                        user->removeAndDeallocate();
+                    }
```
[the remark this answers is not in the document]
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp:1629]: they factually do not hit that same loop, as evidenced by the fact that the trivial example:
```slang
struct Foo
{
    int f;
}

void recursiveB(Foo* i)
{
    recursiveA(i);
}

void recursiveA(Foo* i)
{
    recursiveB(i);
}

export __extern_cpp int main(int argc, NativeString* argv)
{
    Foo* i;
    recursiveA(i);
    return 0;
}
```
does not hang even before this PR, because they do find each other from the cache (duplicating `recursiveA` does not duplicate `recursiveB`) There is nothing more to fix.

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 11]
```diff
@@ -1585,6 +1585,34 @@ struct LoweredElementTypeContext
         auto clonedFunc = as<IRFunc>(cloneInst(&cloneEnv, &builder, call->getCallee()));
         List<IRUse*> uses;
 
+        // If our cloned function has uses already, that means it's calling
```
Comment (csyonghe) [inline source/slang/slang-ir-lower-buffer-element-type.cpp]: I am not sure if this statement is true. If there is a decoration on the function pointing to itself, it doesn't mean the function is recursive.

File: source/slang/slang-ir-lower-buffer-element-type.cpp [hunk 12]
```diff
@@ -1585,6 +1585,34 @@ struct LoweredElementTypeContext
...
+        // itself (= is recursive). Such uses are problematic, as the
+        // `specializedFunc` cache won't contain the newly cloned function,
+        // causing an infinite loop of specialization. The original has already
+        // been handled though, so we can just swap the calls back to calling
+        // the original function and avoid such loops.
```
Comment (csyonghe) [inline source/slang/slang-ir-lower-buffer-element-type.cpp]: is it correct to just call the original function? will that get handled later to call the specialiized func? I don't understand the flow here.

File: source/slang/slang-ir-lower-buffer-element-type.cpp (hunk 11 above)
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp]: That is true... Perhaps I should amend this comment. The whole `if`-gate is basically just an optimization though. It only exists to avoid running code that would do nothing if the function is not recursive.

File: source/slang/slang-ir-lower-buffer-element-type.cpp (hunk 12 above)
Comment (juliusikkala) [reply source/slang/slang-ir-lower-buffer-element-type.cpp]: The flow is indeed a bit tough to figure out. It is pretty closely related to how the `deferStorageToLogicalCasts` function above works. I think that the old behavior here has a couple of separate problems when a function call to itself gets cloned.

The first problem is that the callee becomes the clone function itself, which potentially has a changed type signature, yet the old arguments are retained.

Second, the `CastStorageToLogical` insts in the arguments of the cloned call get recorded in `outNewCasts`, causing the cloned function to get handled as if it had not been specialized yet in here:

https://github.com/juliusikkala/slang/blob/5060a7d0b70e54f6d73bac90d5f4887dfdea462c/source/slang/slang-ir-lower-buffer-element-type.cpp#L1540-L1553

despite the fact that it actually has been specialized already. This causes an endless loop of specializations.

By retaining the call to the original function, we avoid the first issue directly, and the second issue is dodged as the call-to-self still gets inserted in `outNewCasts`, but the next iteration of the loop now finds the function in `specializedFuncs` and replaces the call to the cloned function itself, now with fixed arguments and avoiding the infinite loop.

File: source/slang/slang-ir-lower-buffer-element-type.cpp (hunk 11 above)
Comment (csyonghe) [reply source/slang/slang-ir-lower-buffer-element-type.cpp]: I see, in this case we should update the comment so it doesn't appear as if it's an oversight.

File: source/slang/slang-ir-lower-buffer-element-type.cpp (hunk 12 above)
Comment (csyonghe) [reply source/slang/slang-ir-lower-buffer-element-type.cpp]: I see, that explanation makes sense. Can we add this to the comment explaning why this is correct?

**Review by jkwak-work — APPROVED** [review]
Looks good to me.

It sounds like Yong is fine with the code itself.
The requested comments are improved.
